core/Session/SessionAuth.php

```diff
@@ -0,0 +1,139 @@
+<?php
+/**
+ * Piwik - free/libre analytics platform
+ *
+ * @link http://piwik.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ *
+ */
+
+namespace Piwik\Session;
+
+use Piwik\Auth;
+use Piwik\AuthResult;
+use Piwik\Date;
+use Piwik\Plugins\UsersManager\Model as UsersModel;
+use Piwik\Session;
+
+/**
+ * Validates already authenticated sessions.
+ *
+ * See {@link \Piwik\Session\SessionFingerprint} for more info.
+ */
+class SessionAuth implements Auth
+{
+    /**
+     * For tests, since there's no actual session there.
+     *
+     * @var bool
+     */
+    private $shouldDestroySession;
+
+    public function __construct($shouldDestroySession = true)
+    {
+        $this->shouldDestroySession = $shouldDestroySession;
+    }
+
+    public function getName()
+    {
+        // empty
+    }
+
+    public function setTokenAuth($token_auth)
+    {
+        // empty
+    }
+
+    public function getLogin()
+    {
+        // empty
+    }
+
+    public function getTokenAuthSecret()
+    {
+        // empty
+    }
+
+    public function setLogin($login)
+    {
+        // empty
+    }
+
+    public function setPassword($password)
+    {
+        // empty
+    }
+
+    public function setPasswordHash($passwordHash)
+    {
+        // empty
+    }
+
+    public function authenticate()
+    {
+        $sessionId = new SessionFingerprint();
+        $userModel = new UsersModel();
+
+        $userForSession = $sessionId->getUser();
+        if (empty($userForSession)) {
+            return $this->makeAuthFailure();
+        }
+
+        $user = $userModel->getUser($userForSession);
+        if (empty($user)) {
```

Comment:
I know it is a bit paranoid but lets maybe replace this if by

`if (empty($user) || $user['login'] !== $userForSession) {`

As the authentication is basically based on the `getUser` method... if there ever was a bug somehow in there, it could potentially log you in as someone else.

Reply:
All sanity checks come from legitimate paranoia ;)